[DOWNSTREAM-ONLY] rbd: add support for volume group mirroring

Nikhil-Ladha opened the pull request:
This PR adds the required support volume group mirroring which includes:
- implementing Mirror interface for volume group
- handle replication for both volume group and volume as source
- delete only omap data when the group is secondary
- support for static groups in rbd

NOTE: We need the updated go-ceph for group mirroring, depends on https://github.com/red-hat-storage/go-ceph/pull/1, once it is merged the same will be added to this PR.

This is PR is on top of the changes made in https://github.com/ceph/ceph-csi/pull/4739, credits to @Madhu-1 for the initial PR.

Madhu-1 committed:
rbd: implement volume group using go-ceph

This adds the required functionality to
call the go-ceph API's for the rbd
volume group.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>
Co-authored-by: Nikhil-Ladha <nikhilladha1999@gmail.com>
Signed-off-by: Nikhil-Ladha <nikhilladha1999@gmail.com>

Madhu-1 committed:
rbd: implement GetMirrorSource in manager

implementing GetMirrorSource in manager
to return volume or the volumegroup based
on the replication source, if replication
source is nil return the volume details
for backward compatibility.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>
Co-authored-by: Nikhil-Ladha <nikhilladha1999@gmail.com>
Signed-off-by: Nikhil-Ladha <nikhilladha1999@gmail.com>

Nikhil-Ladha committed:
rbd: delete group only if its primary

delete the group only if it is primary, and when the
group is secondary only delete metadata(OMAP) for the
volumegroup.

Signed-off-by: Nikhil-Ladha <nikhilladha1999@gmail.com>

Nikhil-Ladha committed:
rbd: add support for "dynamic" volume groups

currently, rbd supports only "static" groups.
But, to leverage the functionality of dynamic groups
we are "disabling/enabling" mirroring on the volume
groups when the user performs "add/remove" operations
of PVCs to/from the volume groups.

Signed-off-by: Nikhil-Ladha <nikhilladha1999@gmail.com>

Nikhil-Ladha committed:
vendor: update go-ceph api to red-hat-storage fork

updated go-ceph api to point to red-hat-storage/go-ceph 4.19 branch

Signed-off-by: Nikhil-Ladha <nikhilladha1999@gmail.com>

Madhu-1 commented:
@Nikhil-Ladha can we update the base image so that all the CI passes? or else it will be very difficult see if PR even builds are not afer merging this one.

Madhu-1 commented on internal/csi-addons/rbd/volumegroup.go:
```diff
@@ -403,6 +494,18 @@ func (vs *VolumeGroupServer) ModifyVolumeGroupMembership(
+	}
+	err = mirror.EnableMirroring(ctx, mirroringMode)
+	if err != nil {
+		log.DebugLog(ctx, "failed to enable mirroring after modifying volume group")
```
it should be logged as error

Madhu-1 commented on internal/csi-addons/rbd/volumegroup.go:
```diff
@@ -340,6 +418,23 @@ func (vs *VolumeGroupServer) ModifyVolumeGroupMembership(
+	}
+	err = rbd.DisableVolumeReplication(mirror, ctx, vgrMirrorInfo.IsPrimary(), force)
+	if err != nil {
+		log.DebugLog(ctx, "failed to disable mirroring before modifying volume group")
```
it should be logged as error

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -1081,6 +1081,25 @@ func cleanupRBDImage(ctx context.Context,
 		return nil, status.Errorf(codes.Aborted, "rbd %s is still being used", rbdVol.RbdImageName)
 	}
 
+	// check if the image is part of a group, if yes then error out
+	image, err := rbdVol.open()
+	if err != nil {
+		err = fmt.Errorf("failed to open image %q: %w", rbdVol, err)
+		return &csi.DeleteVolumeResponse{}, err
+	}
+	defer image.Close()
+
+	imageInfo, err := image.GetGroup()
+	if err != nil {
+		err = fmt.Errorf("could not get group information for image %q: %w", rbdVol, err)
+		return &csi.DeleteVolumeResponse{}, err
+	}
+
+	if imageInfo.Name != "" {
+		err = fmt.Errorf("image %q is a part of volume group %q", rbdVol, imageInfo.Name)
+		return &csi.DeleteVolumeResponse{}, err
+	}
+
```
we dont need this check right? RBD is going to fix the problem. If required move this to a help function

Nikhil-Ladha replied:
Oh, yes correct. They should be fixing this issue.

Madhu-1 commented on internal/rbd/group/group_mirror.go:
```diff
@@ -0,0 +1,328 @@
+/*
+Copyright 2025 The Ceph-CSI Authors.
```
Looks like few new lines are missing in the template, can you please check

Nikhil-Ladha commented:
@Madhu-1 @nixpanic the codespell and lint-extras test reports error that are not part of the changes, but the same is not reported upstream, why is that so? Is the ceph image somehow picking up a different versions of the tool and creating a diff?

nixpanic commented:
Added `ci/skip/multi-arch-build` as the only available testing image is for a single architecture.

nixpanic commented:
> @Madhu-1 @nixpanic the codespell and lint-extras test reports error that are not part of the changes, but the same is not reported upstream, why is that so? Is the ceph image somehow picking up a different versions of the tool and creating a diff?

I am not sure about that. The **lint-extras** and **codespell** jobs run in the test-container, which uses `fedora:latest` as base.

Nikhil-Ladha committed:
ci: custom image for testing

custom base_image for testing, commit will be removed
once the PR is good to merge

Signed-off-by: Nikhil-Ladha <nikhilladha1999@gmail.com>

nixpanic commented:
Building fails because some of the librbd functions require an additional parameter

https://github.com/red-hat-storage/go-ceph/blob/a539ad02e45361e3a5e4628af49de704203fe345/rbd/mirror_group.go#L28-L31

![image](https://github.com/user-attachments/assets/15d343bc-99bc-4ac0-92a1-c95fa8a455ba)

Nikhil-Ladha commented:
> Building fails because some of the librbd functions require an additional parameter
> 
> https://github.com/red-hat-storage/go-ceph/blob/a539ad02e45361e3a5e4628af49de704203fe345/rbd/mirror_group.go#L28-L31
> 
> ![image](https://private-user-images.githubusercontent.com/400257/427969967-15d343bc-99bc-4ac0-92a1-c95fa8a455ba.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMxNjIyNDAsIm5iZiI6MTc0MzE2MTk0MCwicGF0aCI6Ii80MDAyNTcvNDI3OTY5OTY3LTE1ZDM0M2JjLTk5YmMtNGFjMC05MmExLWM5NWZhOGE0NTViYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMyOFQxMTM5MDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mYWIzYzI2MjI1NDhjZmVhNWQwMjZhMTk1MWI5ZTAyMjM3ZGJlMzEwOWI2MTJjODljMzQwZDYxNmFhYWRmYTBkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Njj5if394oIzb7UW68yn0jI9F5QoufCZFsnsQcM5v2I)

Yeah, it seems like I was using the wrong build number to test 😶‍🌫️ 
The tests looks good now, the `lint-extras` and `codespell` issues should be fixed when we have a stable 4.19 build and the fix is backported to the 4.19 branch

nixpanic commented on internal/csi-addons/rbd/replication.go:
```diff
@@ -29,11 +29,13 @@ import (
 	"github.com/ceph/ceph-csi/internal/util/log"
 
 	librbd "github.com/ceph/go-ceph/rbd"
+
```
no need for this empty line

nixpanic commented on internal/csi-addons/rbd/replication.go:
```diff
@@ -29,11 +29,13 @@ import (
 	csicommon "github.com/ceph/ceph-csi/internal/csi-common"
 	"github.com/ceph/ceph-csi/internal/rbd"
 	corerbd "github.com/ceph/ceph-csi/internal/rbd"
+	rbd_group "github.com/ceph/ceph-csi/internal/rbd/group"
```
Why not just stick to `group`? Ideally there should not be a need to import this, as the `internal/rbd/types` should provide a complete interface for VolumeGroup operations.

nixpanic commented on internal/csi-addons/rbd/replication.go:
```diff
@@ -785,13 +787,14 @@ func getGRPCError(err error) error {
+		corerbd.ErrAborted:               codes.Aborted,
+		corerbd.ErrFailedPrecondition:    codes.FailedPrecondition,
+		corerbd.ErrUnavailable:           codes.Unavailable,
+		rbd_group.ErrRBDGroupUnAvailable: codes.Unavailable,
```
this error should be part of the new `internal/rbd/errors` package

nixpanic commented on internal/rbd/group/group_mirror.go:
```diff
@@ -0,0 +1,331 @@
+	"github.com/ceph/ceph-csi/internal/util/log"
+)
+
+var ErrRBDGroupUnAvailable = errors.New("RBD group is unavailable")
```
no need to include `RBD` in the name, this is part of an `rbd` package already. Move the error to `internal/rbd/errors`. Usually it is written like `Unavailable`, not `UnAvailable`.

nixpanic commented on internal/rbd/group/group_mirror.go:
```diff
@@ -0,0 +1,331 @@
+
+var ErrRBDGroupUnAvailable = errors.New("RBD group is unavailable")
+
+type volumeGroupMirror struct {
```
There is nothing added to the struct, so I guess you can use the simpler `type volumeGroupMirror volumeGroup` too?

nixpanic commented on internal/rbd/group/group_mirror.go:
```diff
@@ -0,0 +1,331 @@
+
+// groupInfo is a wrapper around librbd.MirrorGroupInfo that contains the
+// group mirror info.
+type groupInfo struct {
```
can this be written like `type groupInfo librbd.MirrorGroupInfo`?

Nikhil-Ladha replied:
We can, but since go-ceph APIs returns a librbd type, we would have to do a type cast before returning, won't that be a bit messy?
Also, this coding style keeps it consistent with the `mirror.go` struct definitions for the same type.
_(Applies to all the other type concerns, you mentioned below)_

nixpanic replied:
I think you can just write is like
```go
mgi := librbd.SomeMirrorGroupInfoFunc()
gi := groupInfo(mgi)
```

That is surely shorter and simpler than
```go
mgi := librbd.SomeMirrorGroupInfoFunc()
gi := groupInfo{MirrorGroupInfo: mgi}
```

The 2nd format gives the suggestion that `groupInfo` has other attributes that can (or may need to?) be set.

nixpanic commented on internal/rbd/group/group_mirror.go:
```diff
@@ -0,0 +1,331 @@
+	*volumeGroup
+}
+
+func (vg volumeGroupMirror) EnableMirroring(ctx context.Context, mode librbd.ImageMirrorMode) error {
```
Strange, I would expect `func (vg *volumeGroupMirror) ...`, with a pointer to `volumeGroupMirror`.

Better call this `vgm` instead of `vg` to make it clear this is a different type.

nixpanic commented on internal/rbd/group/group_mirror.go:
```diff
@@ -0,0 +1,331 @@
+
+// globalMirrorGroupStatus is a wrapper around librbd.GlobalGroupMirrorImageStatus that contains the
+// global mirror group status.
+type globalMirrorGroupStatus struct {
```
same here, if the struct does not include anything extra, just use `type globalMirrorGroupStatus librbd.GlobalMirrorGroupStatus`.

nixpanic commented on internal/rbd/group/group_mirror.go:
```diff
@@ -0,0 +1,331 @@
+	librbd.GlobalMirrorGroupStatus
+}
+
+func (status *globalMirrorGroupStatus) GetState() string {
```
These `groupInfo` and `globalMirrorGroupStatus` functions look simple enough to have unit-tests. It would be good to add those.

nixpanic commented on internal/rbd/group/group_mirror.go:
```diff
@@ -0,0 +1,331 @@
+func (status *globalMirrorGroupStatus) GetRemoteSiteStatus(ctx context.Context) (types.SiteStatus, error) {
+	var (
+		ss  librbd.SiteMirrorGroupStatus
+		err error = librbd.ErrNotExist
```
Use a ceph-csi error instead of a librbd one?

nixpanic commented on internal/rbd/group/group_mirror.go:
```diff
@@ -0,0 +1,331 @@
+
+// siteMirrorGroupStatus is a wrapper around librbd.SiteMirrorGroupStatus that contains the
+// site mirror group status.
+type siteMirrorGroupStatus struct {
```
same `type` concern

nixpanic commented on internal/csi-addons/rbd/replication.go:
```diff
@@ -272,24 +272,23 @@ func (rs *ReplicationServer) EnableVolumeReplication(ctx context.Context,
 	defer mgr.Destroy(ctx)
 
-	rbdVol, err := mgr.GetVolumeByID(ctx, volumeID)
+	volumes, mirror, err := mgr.GetMirrorSource(ctx, reqID, req.GetReplicationSource())
```
should `mirror` not be destroyed as well?

Nikhil-Ladha replied:
We are destroying the volumes/volumeGroup, and mirror is just an extension of the same and doesn't use a separate connection or resource. So, no we don't need to destroy mirror explicitly IIUC.

nixpanic commented on internal/csi-addons/rbd/replication.go:
```diff
@@ -332,34 +342,33 @@ func (rs *ReplicationServer) EnableVolumeReplication(ctx context.Context,
 	defer mgr.Destroy(ctx)
 
-	rbdVol, err := mgr.GetVolumeByID(ctx, volumeID)
+	volumes, mirror, err := mgr.GetMirrorSource(ctx, reqID, req.GetReplicationSource())
```
should `mirror` not be destroyed as well? And in more places...

nixpanic commented on internal/csi-addons/rbd/replication.go:
```diff
@@ -695,35 +704,40 @@ func (rs *ReplicationServer) ResyncVolume(ctx context.Context,
-		if sErr != nil {
-			return nil, status.Errorf(codes.Internal, "failed to parse image creation time: %s", sErr.Error())
+		// image creation time is stored in the image metadata. it looks like
+		// `"seconds:1692879841 nanos:631526669"`
```
no need to mention the format if it is not parsed here (should be parsed in an other function anyway)

Nikhil-Ladha replied:
It is an old comment, it was not added as part of this PR.
The diff is because of the for loop added around the code.

nixpanic commented on internal/rbd/types/group.go:
```diff
@@ -40,6 +40,8 @@ type journalledObject interface {
 
 // VolumeGroup contains a number of volumes.
+//
+//nolint:interfacebloat // VolumeGroup has more than 10 methods, that is ok.
```
I do not think "that is ok" :wink:  In upstream the `Destroy()` function moved to `journalledObject` already, maybe include that commit in this PR too?

nixpanic commented on internal/csi-addons/rbd/volumegroup.go:
```diff
@@ -206,27 +207,40 @@ func (vs *VolumeGroupServer) DeleteVolumeGroup(
 
-	// verify that the volume group is empty
-	volumes, err := vg.ListVolumes(ctx)
+	_, mirror, err := mgr.GetMirrorSource(ctx, req.GetVolumeGroupId(), &replication.ReplicationSource{
```
What about the defer Destroy() for all the volumes (`_`) that were returned? This probably leaks resources now.

nixpanic commented on internal/csi-addons/rbd/volumegroup.go:
```diff
@@ -295,6 +309,25 @@ func (vs *VolumeGroupServer) ModifyVolumeGroupMembership(
+		return nil, getGRPCError(err)
+	}
+
+	vgrMirrorInfo, err := mirror.GetMirroringInfo(ctx)
```
need to check `err`?

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -1081,6 +1081,14 @@ func cleanupRBDImage(ctx context.Context,
 	}
 
+	// check if image is part of a group
+	err = isVolumePartOfGroup(rbdVol)
```
use `rbdVol.GetVolumeGroupID()`

Nikhil-Ladha replied:
Done

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -1109,6 +1117,32 @@ func cleanupRBDImage(ctx context.Context,
 }
 
+// isVolumePartOfGroup checks if the image is part of a group.
+func isVolumePartOfGroup(rbdVol *rbdVolume) error {
```
drop this in favor of the upstream `Volume.GetVolumeGroupID()` function

https://github.com/ceph/ceph-csi/blob/9002d95e5321d73609c552d160cc9361c5279075/internal/rbd/types/volume.go#L42

nixpanic commented on internal/rbd/group/volume_group.go:
```diff
@@ -26,15 +26,18 @@ import (
-	ErrRBDGroupNotFound     = fmt.Errorf("%w: RBD group not found", librbd.ErrNotFound)
+	ErrRBDGroupNotConnected    = fmt.Errorf("%w: RBD group is not connected", librados.ErrNotConnected)
+	ErrRBDGroupNotFound        = fmt.Errorf("%w: RBD group not found", librbd.ErrNotFound)
+	ErrRBDGroupInvalidArgument = errors.New("RBD group invalid arguments provided")
```
move the to `internal/rbd/errors`

nixpanic commented on internal/rbd/group/volume_group.go:
```diff
@@ -184,7 +187,7 @@ func (vg *volumeGroup) Create(ctx context.Context) error {
 }
 
-func (vg *volumeGroup) Delete(ctx context.Context) error {
+func (vg *volumeGroup) Delete(ctx context.Context, vgrMirrorInfo types.MirrorInfo, mirror types.Mirror) error {
```
ideally this should just stay a plain `Delete()` function. If something fancy like this needed, add a `types.Mirror.Delete()` function with more parameters.

Nikhil-Ladha replied:
Ack, updated the code to fetch the required params in the function itself.

nixpanic commented on internal/rbd/group/volume_group.go:
```diff
@@ -258,13 +300,13 @@ func (vg *volumeGroup) RemoveVolume(ctx context.Context, vol types.Volume) error
 	if err != nil {
-		if errors.Is(err, librbd.ErrNotExist) {
-			return nil
+		if !errors.Is(err, librbd.ErrNotExist) {
```
Why change the logic? If it is needed to continue with removing the volume if it is not part of the group, do this in a separate (upstream) commit.

Nikhil-Ladha replied:
Yes, we should continue with the removal of image even if the image is not part of group/doesn't exist in group.
I will send a PR for this upstream.

Nikhil-Ladha replied:
Here's the PR for the same: https://github.com/ceph/ceph-csi/pull/5250

nixpanic commented on internal/rbd/types/group.go:
```diff
@@ -60,7 +60,7 @@ type VolumeGroup interface {
 
 	// Delete removes the VolumeGroup from the backend storage.
-	Delete(ctx context.Context) error
+	Delete(ctx context.Context, vgMirrorInfo MirrorInfo, mirror Mirror) error
```
the simple VolumeGroup interface should not need to bother with anything else than basic operations, don't add these parameters here (maybe in the `Mirror` type?)

nixpanic commented on internal/csi-addons/rbd/volumegroup.go:
```diff
@@ -22,16 +22,17 @@ import (
+	"github.com/ceph/ceph-csi/internal/rbd"
+	"github.com/ceph/ceph-csi/internal/rbd/group"
+	"github.com/ceph/ceph-csi/internal/rbd/types"
+	"github.com/ceph/ceph-csi/internal/util/log"
```
ceph-csi prefers local imports before 3rd party ones (for whatever unknown to me reason). This is an unrelated change that should be dropped.

nixpanic commented on internal/csi-addons/rbd/volumegroup.go:
```diff
@@ -207,7 +208,7 @@ func (vs *VolumeGroupServer) DeleteVolumeGroup(
 	log.DebugLog(ctx, "VolumeGroup %q has been found", req.GetVolumeGroupId())
 
-	_, mirror, err := mgr.GetMirrorSource(ctx, req.GetVolumeGroupId(), &replication.ReplicationSource{
+	volumes, mirror, err := mgr.GetMirrorSource(ctx, req.GetVolumeGroupId(), &replication.ReplicationSource{
```
need to destroy the volumes and mirror?

nixpanic reviewed (changes requested):
I've reviewed each commit separately, some comments were addressed in follow-up commits, it seems. Needs a little bit more work, specially also because #508 will probably be backported to the release-4.19 branch and will cause conflicts.

Nikhil-Ladha commented:
@nixpanic addressed most of the reviews, but some are still left like using the functions that were recently pushed upstream. I am waiting for a stable build to happen for 4.19 and once that is done, I will rebase the PR on top of those changes and address those comments as well.

Nikhil-Ladha commented:
The PR is rebased on top of latest code in upstream and the 4.19 branch, so please do a second round of review.
Thanks.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -1082,6 +1086,21 @@ func cleanupRBDImage(ctx context.Context,
+	if err != nil && !errors.Is(err, rbderrors.ErrGroupNotFound) {
+		log.ErrorLog(ctx, err.Error())
+
+		return nil, err
```
return GRPC error

Nikhil-Ladha replied:
Done

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -1082,6 +1086,21 @@ func cleanupRBDImage(ctx context.Context,
+		err = fmt.Errorf("image %q is a part of volume group %q", rbdVol, groupId)
+		log.ErrorLog(ctx, err.Error())
+
+		return nil, err
```
return GRPC error

Nikhil-Ladha replied:
Done

Rakshith-R commented on internal/rbd/errors/errors.go:
```diff
@@ -67,4 +67,8 @@ var (
 	ErrGroupNotConnected = fmt.Errorf("%w: RBD group is not connected", librados.ErrNotConnected)
 	// ErrGroupNotFound is returned when group is not found in the cluster on the given pool and/or namespace.
 	ErrGroupNotFound = fmt.Errorf("%w: RBD group not found", librbd.ErrNotFound)
+	// ErrGroupUnavailable is returned when the requested group is unavailable.
+	ErrGroupUnavailable = errors.New("rbd group is unavailable")
```
Can you add more info here

similar to the info documented for ErrUnavailable above
```
	// ErrUnavailable is returned when the image needs to be recreated
	// locally and may be corrected by retrying with a backoff.
	ErrUnavailable = errors.New("image needs to be recreated")
```

maybe keep the error message very similar too

Nikhil-Ladha replied:
Done

Rakshith-R commented on internal/rbd/group/volume_group.go:
```diff
@@ -190,6 +192,57 @@ func (vg *volumeGroup) Delete(ctx context.Context) error {
+			if localStatus.IsUP() && localStatus.GetState() == librbd.MirrorGroupStatusStateReplaying.String() {
+				return vg.commonVolumeGroup.Delete(ctx)
+			}
+			log.ErrorLog(ctx,
+				"%w: secondary group status is up=%t and state=%s",
+				rbderrors.ErrGroupInvalidArgument,
+				localStatus.IsUP(),
+				localStatus.GetState())
+		}
+	}
```
return statement is missing here ?

Nikhil-Ladha replied:
Done

Rakshith-R commented on internal/csi-addons/rbd/volumegroup.go:
```diff
@@ -527,6 +597,21 @@ func (vs *VolumeGroupServer) ModifyVolumeGroupMembership(
 		}
 	}
 
+	// Enable mirroring after modification of volume group
+	if isMirroringEnabled {
```
`isMirroringEnabled` is a local variable

what if there is a restart during a modifymembership call with change in PVCs in group ? 

Is that scenario tested ? 
Does it go smoothly ?

Madhu-1 replied:
This is not global variable its per RPC call, Do you see any problem with it?

Rakshith-R replied:
> This is not global variable its per RPC call, Do you see any problem with it?

My questions are mentioned in the first comment.

The expectation form the variable is that if mirroring was enabled before modifymembership call, mirroring will be switched back on at the end of modification.
If there is a restart, when the new modify call begins, mirroring will not be enabled, thus mirroring will not get switched back on at the end.

Madhu-1 replied:
Yes, This is for dynamic grouping, CSI will first disable mirroring add images to the group and reenable the mirroring. If the mirroring is already disabled it will just add the images to the group and enable the mirroring again.

Madhu-1 replied:
got it, Thanks for catching this. i miss read it.

Nikhil-Ladha replied:
Updated the code to remove the check from enable mirroring completely.
We should not bother checking on enable mirroring, as it will return success even if it's enabled.
So, we should be good with this.

Madhu-1 commented on internal/csi-addons/rbd/volumegroup.go:
```diff
@@ -527,6 +597,21 @@ func (vs *VolumeGroupServer) ModifyVolumeGroupMembership(
 		}
 	}
 
+	// Enable mirroring after modification of volume group
```
i think the check need to be changed to `!` do enable mirroring if its disabled.

nixpanic commented on internal/csi-addons/rbd/replication.go:
```diff
@@ -308,17 +307,28 @@ func (rs *ReplicationServer) EnableVolumeReplication(ctx context.Context,
+					err = fmt.Errorf("failed to handle parent image for volume group %q: %w", mirror, err)
+					return nil, getGRPCError(err)
+				}
+				err = mirror.EnableMirroring(ctx, mirroringMode)
```
there are a lot of calls to this, why not combine it with the one of the `else` below?

nixpanic commented on internal/csi-addons/rbd/volumegroup.go:
```diff
@@ -421,10 +426,55 @@ func (vs *VolumeGroupServer) ModifyVolumeGroupMembership(
+	// If the mirroring is not enabled and we are modifying the group, that means
+	// we are doing a cleanup and we shouldn't be checking for mirror status
+	isMirroringEnabled := false
+	if vgrMirrorInfo.GetState() == librbd.MirrorImageEnabled.String() {
```
This part of the function is really quite complex, it can benefit a lot from splitting this out in one or more helper functions. Specially the `librbd` checking should be done somewhere in the `internal/rbd` package, the `internal/csi-addons/rbd` package should only use go-ceph indirectly. However, it seems that `internal/csi-addons/rbd/replication.go` violates this rule already too :disappointed: You can keep this for now, but it does need improving before it can be accepted in upstream.

Nikhil-Ladha replied:
Once, we have dynamic grouping available in ceph. We can remove all these checks, the checks are in place only to support dynamic grouping and locking issues in ceph rbd.

nixpanic commented on scripts/Dockerfile.test:
```diff
@@ -7,8 +7,7 @@
 #
-
-FROM quay.io/fedora/fedora:latest
+FROM quay.ceph.io/ceph-ci/ceph:wip-pkalever-rbd-group-snap-mirror
```
Isn't this getting garbage collected? I thought there was a copy somewhere that would stay around longer?

Nikhil-Ladha replied:
Not anytime soon, at least not till the upstream PR is merged.

Madhu-1 commented:
/lgtm
/approve
